File: cracking_the_coding_interview/chapter_1/unique_chars1.py
# Interview Question 1.1

# Time complexity: O(n)
# Auxiliary space complexity: O(n)

def unique_chars1(s):
    return len(s) == len(set(s))


if __name__ == '__main__':
    print(unique_chars1(""))
    print(unique_chars1("a"))
    print(unique_chars1("aa"))
    print(unique_chars1("abc"))
    print(unique_chars1("abca"))
    print(unique_chars1("abcb"))
    print(unique_chars1("abcc"))

File: cracking_the_coding_interview/chapter_1/UniqueChars1.java
// Interview Question 1.1

// Time complexity: O(n)
// Auxiliary space complexity: O(n)

import java.util.Set;
import java.util.HashSet;

public class UniqueChars1 {

    public static boolean uniqueChars1(String str) {
        Set<Character> set = new HashSet<Character>();

        for (int i = 0; i < str.length(); i++) {
            char c = str.charAt(i);
            if (set.contains(c)) {
                return false;
            }
            set.add(c);
        }

        return true;
    }

    public static void main(String[] args) {
        System.out.println(uniqueChars1(""));
        System.out.println(uniqueChars1("a"));
        System.out.println(uniqueChars1("aa"));
        System.out.println(uniqueChars1("abc"));
        System.out.println(uniqueChars1("abca"));
        System.out.println(uniqueChars1("abcb"));
        System.out.println(uniqueChars1("abcc"));
    }
}
